src/esp/io/URDFParser.h

```diff
@@ -0,0 +1,314 @@
+};
+
+struct UrdfModel {
+  std::string m_name;
+  std::string m_sourceFile;
+  Magnum::Matrix4 m_rootTransformInWorld(Magnum::Math::IdentityInitT);
+
+  //! map of names to materials
+  std::map<std::string, std::shared_ptr<UrdfMaterial>> m_materials;
+
+  //! map of names to links
+  std::map<std::string, std::shared_ptr<UrdfLink>> m_links;
+
+  //! map of link indices to names
+  std::map<int, std::string> m_linkIndicesToNames;
+
+  //! map of names to joints
+  std::map<std::string, std::shared_ptr<UrdfJoint>> m_joints;
+
+  //! list of root links (usually 1)
+  std::vector<std::shared_ptr<UrdfLink>> m_rootLinks;
+  bool m_overrideFixedBase;
+
+  void printKinematicChain() const;
+
+  std::shared_ptr<UrdfLink> getLink(std::string linkName) const {
+    if (m_links.count(linkName)) {
+      return m_links.at(linkName);
+    }
+    return nullptr;
+  }
+  std::shared_ptr<UrdfLink> getLink(int linkIndex) const {
+    if (m_linkIndicesToNames.count(linkIndex)) {
+      return getLink(m_linkIndicesToNames.at(linkIndex));
+    }
+    return nullptr;
+  }
+
+  //! get the parent joint of a link
+  std::shared_ptr<UrdfJoint> getJoint(int linkIndex) const {
+    if (m_linkIndicesToNames.count(linkIndex)) {
+      return getLink(m_linkIndicesToNames.at(linkIndex))->m_parentJoint;
+    }
+    return nullptr;
+  }
+
+  UrdfModel() : m_overrideFixedBase(false) {}
+};
+
+class URDFParser {
+  // datastructures
+  UrdfModel m_urdfModel;
+  float m_urdfScaling = 1.0;
+
+  // URDF file path of last load call
+  std::string sourceFilePath_;
+
+  // parser functions
+  bool parseTransform(Magnum::Matrix4& tr, tinyxml2::XMLElement* xml);
+  bool parseInertia(UrdfInertia& inertia, tinyxml2::XMLElement* config);
+  bool parseGeometry(UrdfGeometry& geom, tinyxml2::XMLElement* g);
+  bool parseVisual(UrdfModel& model,
+                   UrdfVisual& visual,
+                   tinyxml2::XMLElement* config);
+  bool parseCollision(UrdfCollision& collision, tinyxml2::XMLElement* config);
+  bool initTreeAndRoot(UrdfModel& model);
+  bool parseMaterial(UrdfMaterial& material, tinyxml2::XMLElement* config);
+  bool parseJointLimits(UrdfJoint& joint, tinyxml2::XMLElement* config);
+  bool parseJointDynamics(UrdfJoint& joint, tinyxml2::XMLElement* config);
+  bool parseJoint(UrdfJoint& joint, tinyxml2::XMLElement* config);
+  bool parseLink(UrdfModel& model,
+                 UrdfLink& link,
+                 tinyxml2::XMLElement* config);
+  bool parseSensor(UrdfModel& model,
+                   UrdfLink& link,
+                   UrdfJoint& joint,
+                   tinyxml2::XMLElement* config);
+
+  bool validateMeshFile(std::string& filename);
+
+ public:
+  URDFParser(){};
+
+  // parse a loaded URDF string into relevant general data structures
+  // return false if the string is not a valid urdf or other error causes abort
+  bool parseURDF(const std::string& meshFilename);
+
+  void setGlobalScaling(float scaling) { m_urdfScaling = scaling; }
```

Comment:
This API is a little confusing. If I call `parseURDF`, then `setGlobalScaling`, then `getModel`, do I get a scaled model?